File: src/linear_algebra_matrix_inverses_medium.py
"""
linear_algebra - matrix_inverses (medium)
Generated: 2026-02-22T05:51:15.984753
"""

from problem_utils import *

def generate():
    problem_type = choice(['2x2_inverse', '2x2_determinant_check', '3x3_inverse', 'verify_inverse'])
    
    if problem_type == '2x2_inverse':
        # Generate a 2x2 invertible matrix
        # Start with integer entries that give a nice inverse
        det = choice([-2, -1, 1, 2])
        a_val = nonzero(-3, 3)
        b_val = randint(-3, 3)
        c_val = randint(-3, 3)
        # d chosen so determinant = det
        d_val = (det + a_val * c_val) // a_val if a_val != 0 else nonzero(-3, 3)
        
        # Rebuild to ensure integer determinant
        a_val = nonzero(-3, 3)
        d_val = nonzero(-3, 3)
        b_val = randint(-2, 2)
        c_val = randint(-2, 2)
        
        # Check det != 0
        det = a_val * d_val - b_val * c_val
        if det == 0:
            d_val += 1
            det = a_val * d_val - b_val * c_val
        
        M = Matrix([[a_val, b_val], [c_val, d_val]])
        M_inv = M.inv()
        
        return problem(
            question=f"Find the inverse of the matrix $A = {latex(M)}$",
            answer=M_inv,
            difficulty=(1300, 1450),
            topic="linear_algebra/matrix_inverses",
            solution=steps(
                f"For a $2 \\times 2$ matrix $A = \\begin{{pmatrix}} a & b \\\\ c & d \\end{{pmatrix}}$, the inverse is $A^{{-1}} = \\frac{{1}}{{ad-bc}} \\begin{{pmatrix}} d & -b \\\\ -c & a \\end{{pmatrix}}$",
                f"Calculate the determinant: $\\det(A) = ({a_val})({d_val}) - ({b_val})({c_val}) = {det}$",
                f"Since $\\det(A) \\neq 0$, the matrix is invertible",
                f"$A^{{-1}} = \\frac{{1}}{{{det}}} \\begin{{pmatrix}} {d_val} & {-b_val} \\\\ {-c_val} & {a_val} \\end{{pmatrix}} = {latex(M_inv)}$"
            )
        )
    
    elif problem_type == '2x2_determinant_check':
        # Give a singular matrix and ask if it's invertible
        a_val = nonzero(-4, 4)
        b_val = nonzero(-4, 4)
        # Make it singular: set c, d so det = 0
        c_val = nonzero(-3, 3)
        d_val = (b_val * c_val) // a_val if a_val != 0 else b_val
        
        M = Matrix([[a_val, b_val], [c_val, d_val]])
        det = M.det()
        
        if det == 0:
            ans = False
            inv_text = "not invertible"
        else:
            ans = True
            inv_text = "invertible"
        
        return problem(
            question=f"Is the matrix $A = {latex(M)}$ invertible? (Answer true or false)",
            answer=ans,
            difficulty=(1300, 1400),
            topic="linear_algebra/matrix_inverses",
            solution=steps(
                f"A matrix is invertible if and only if its determinant is nonzero",
                f"Calculate $\\det(A) = ({a_val})({d_val}) - ({b_val})({c_val}) = {det}$",
                f"Since $\\det(A) = {det}$, the matrix is {inv_text}"
            )
        )
    
    elif problem_type == '3x3_inverse':
        # Generate a simple 3x3 invertible matrix
        # Use a diagonal or nearly diagonal matrix for cleaner arithmetic
        diag_type = choice(['diagonal', 'upper_triangular'])
        
        if diag_type == 'diagonal':
            d1, d2, d3 = nonzero(-3, 3), nonzero(-3, 3), nonzero(-3, 3)
            M = Matrix([[d1, 0, 0], [0, d2, 0], [0, 0, d3]])
            M_inv = M.inv()
            
            return problem(
                question=f"Find the inverse of the matrix $A = {latex(M)}$",
                answer=M_inv,
                difficulty=(1350, 1500),
                topic="linear_algebra/matrix_inverses",
                solution=steps(
                    f"This is a diagonal matrix",
                    f"The inverse of a diagonal matrix is obtained by inverting each diagonal entry",
                    f"$A^{{-1}} = {latex(M_inv)}$"
                )
            )
        else:
            # Upper triangular with 1s on diagonal for easier inverse
            a_val = randint(-2, 2)
            b_val = randint(-2, 2)
            c_val = randint(-2, 2)
            M = Matrix([[1, a_val, b_val], [0, 1, c_val], [0, 0, 1]])
            M_inv = M.inv()
            
            return problem(
                question=f"Find the inverse of the matrix $A = {latex(M)}$",
                answer=M_inv,
                difficulty=(1450, 1600),
                topic="linear_algebra/matrix_inverses",
                solution=steps(
                    f"This is an upper triangular matrix with 1s on the diagonal",
                    f"Calculate determinant: $\\det(A) = 1 \\cdot 1 \\cdot 1 = 1 \\neq 0$, so it's invertible",
                    f"Using row reduction or the formula for matrix inverse:",
                    f"$A^{{-1}} = {latex(M_inv)}$"
                )
            )
    
    else:  # verify_inverse
        # Give two matrices and ask to verify if they are inverses
        a_val = nonzero(-2, 2)
        d_val = nonzero(-2, 2)
        b_val = choice([-1, 0, 1])
        c_val = choice([-1, 0, 1])
        
        det = a_val * d_val - b_val * c_val
        if det == 0:
            d_val += 1
            det = a_val * d_val - b_val * c_val
        
        M = Matrix([[a_val, b_val], [c_val, d_val]])
        
        if choice([True, False]):
            # Give the correct inverse
            M_inv = M.inv()
            product = M * M_inv
            ans = True
            are_text = "are"
        else:
            # Give a wrong matrix
            M_inv = Matrix([[d_val, -b_val + 1], [-c_val, a_val]])
            product = M * M_inv
            ans = (product == eye(2))
            are_text = "are" if ans else "are not"
        
        return problem(
            question=f"Verify whether $B = {latex(M_inv)}$ is the inverse of $A = {latex(M)}$ (Answer true or false)",
            answer=ans,
            difficulty=(1350, 1500),
            topic="linear_algebra/matrix_inverses",
            solution=steps(
                f"Two matrices are inverses if their product is the identity matrix",
                f"Calculate $AB = {latex(M)} {latex(M_inv)} = {latex(product)}$",
                f"Since $AB = {latex(product)}$, they {are_text} inverses"
            )
        )

emit(generate())

File: src/linear_algebra_matrix_inverses_medium.jl
# linear_algebra - matrix_inverses (medium)
# Generated: 2026-03-08T21:08:52.715769

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("linear_algebra/matrix_inverses")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # 2x2 matrix inverse computation
        a = nonzero(-5, 5)
        b = randint(-5, 5)
        c = randint(-5, 5)
        d = nonzero(-5, 5)
        
        det_val = a*d - b*c
        for _attempt in 1:1000
            det_val == 0 || break
            d = nonzero(-5, 5)
            det_val = a*d - b*c
        end

        inv_a = d // det_val
        inv_b = -b // det_val
        inv_c = -c // det_val
        inv_d = a // det_val
        
        A_str = "\\begin{bmatrix} $a & $b \\\\ $c & $d \\end{bmatrix}"
        inv_str = "\\begin{bmatrix} $(inv_a) & $(inv_b) \\\\ $(inv_c) & $(inv_d) \\end{bmatrix}"
        
        problem(
            question="Find the inverse of the matrix \$A = $A_str\$",
            answer=inv_str,
            difficulty=(1200, 1400),
            answer_type="matrix",
            solution=steps(
                "Given: \$A = $A_str\$",
                sol("Determinant", "\\det(A) = $(a)($(d)) - $(b)($(c)) = $(det_val)"),
                "Apply formula: \$A^{-1} = \\frac{1}{\\det(A)} \\begin{bmatrix} d & -b \\\\ -c & a \\end{bmatrix}\$",
                sol("Answer", inv_str)
            ),
            time=120
        )
        
    elseif problem_type == 2
        # Check if matrix is invertible (determinant test)
        a = randint(-6, 6)
        b = randint(-6, 6)
        c = randint(-6, 6)
        
        if choice([true, false])
            # Make it non-invertible (det = 0)
            d = (b * c) // a
            if denominator(d) != 1
                d = 0
            else
                d = numerator(d)
            end
            ans = "No"
            det_val = a*d - b*c
        else
            # Make it invertible
            d = nonzero(-6, 6)
            det_val = a*d - b*c
            for _attempt in 1:1000
                det_val == 0 || break
                d = nonzero(-6, 6)
                det_val = a*d - b*c
            end
            ans = "Yes"
        end
        
        A_str = "\\begin{bmatrix} $a & $b \\\\ $c & $d \\end{bmatrix}"
        
        problem(
            question="Is the matrix \$A = $A_str\$ invertible? Answer Yes or No.",
            answer=ans,
            difficulty=(1100, 1300),
            answer_type="text",
            solution=steps(
                "Given: \$A = $A_str\$",
                sol("Determinant", "\\det(A) = $(a)($(d)) - $(b)($(c)) = $(det_val)"),
                "A matrix is invertible if and only if its determinant is nonzero",
                sol("Answer", ans)
            ),
            time=90
        )
        
    elseif problem_type == 3
        # Solve system using matrix inverse: Ax = b
        a = nonzero(-4, 4)
        b = randint(-4, 4)
        c = randint(-4, 4)
        d = nonzero(-4, 4)
        
        det_val = a*d - b*c
        for _attempt in 1:1000
            det_val == 0 || break
            d = nonzero(-4, 4)
            det_val = a*d - b*c
        end

        # Choose solution
        x_sol = randint(-8, 8)
        y_sol = randint(-8, 8)
        
        # Compute b vector
        e = a*x_sol + b*y_sol
        f = c*x_sol + d*y_sol
        
        A_str = "\\begin{bmatrix} $a & $b \\\\ $c & $d \\end{bmatrix}"
        b_str = "\\begin{bmatrix} $e \\\\ $f \\end{bmatrix}"
        ans_str = "($x_sol, $y_sol)"
        
        problem(
            question="Solve the system \$A\\mathbf{x} = \\mathbf{b}\$ where \$A = $A_str\$ and \$\\mathbf{b} = $b_str\$. Give your answer as an ordered pair \$(x, y)\$.",
            answer=ans_str,
            difficulty=(1400, 1600),
            answer_type="point",
            solution=steps(
                "Given: \$A = $A_str\$, \$\\mathbf{b} = $b_str\$",
                sol("Determinant", "\\det(A) = $(det_val)"),
                "Compute \$A^{-1}\$ and multiply: \$\\mathbf{x} = A^{-1}\\mathbf{b}\$",
                sol("Answer", "(x, y) = $ans_str")
            ),
            time=180
        )
        
    elseif problem_type == 4
        # Find determinant given that inverse exists and one element
        a = nonzero(-5, 5)
        b = randint(-5, 5)
        c = randint(-5, 5)
        d = nonzero(-5, 5)
        
        det_val = a*d - b*c
        for _attempt in 1:1000
            (det_val == 0 || abs(det_val) > 20) || break
            d = nonzero(-5, 5)
            det_val = a*d - b*c
        end
        
        inv_11 = d // det_val
        
        A_str = "\\begin{bmatrix} $a & $b \\\\ $c & $d \\end{bmatrix}"
        
        problem(
            question="If \$A = $A_str\$ and \$A^{-1} = \\begin{bmatrix} $(inv_11) & * \\\\ * & * \\end{bmatrix}\$ (where * represents unknown entries), find \$\\det(A)\$.",
            answer=det_val,
            difficulty=(1300, 1500),
            solution=steps(
                "Given: \$A = $A_str\$",
                "The (1,1) entry of \$A^{-1}\$ is \$\\frac{d}{\\det(A)} = \\frac{$d}{\\det(A)} = $(inv_11)\$",
                "Solving: \$\\det(A) = \\frac{$d}{$(inv_11)} = $(det_val)\$",
                sol("Answer", det_val)
            ),
            time=150
        )
        
    else
        # Verify inverse relationship: check if AB = I
        a = nonzero(-3, 3)
        b = randint(-3, 3)
        c = randint(-3, 3)
        d = nonzero(-3, 3)
        
        det_val = a*d - b*c
        for _attempt in 1:1000
            det_val == 0 || break
            d = nonzero(-3, 3)
            det_val = a*d - b*c
        end

        inv_a = d // det_val
        inv_b = -b // det_val
        inv_c = -c // det_val
        inv_d = a // det_val
        
        # Multiply to verify we get I
        prod_11 = a*inv_a + b*inv_c
        prod_12 = a*inv_b + b*inv_d
        prod_21 = c*inv_a + d*inv_c
        prod_22 = c*inv_b + d*inv_d
        
        A_str = "\\begin{bmatrix} $a & $b \\\\ $c & $d \\end{bmatrix}"
        B_str = "\\begin{bmatrix} $(inv_a) & $(inv_b) \\\\ $(inv_c) & $(inv_d) \\end{bmatrix}"
        
        ans = "Yes"
        
        problem(
            question="Verify that \$B = $B_str\$ is the inverse of \$A = $A_str\$ by computing \$AB\$. Does \$AB = I\$? Answer Yes or No.",
            answer=ans,
            difficulty=(1300, 1500),
            answer_type="text",
            solution=steps(
                "Given: \$A = $A_str\$, \$B = $B_str\$",
                "Compute: \$AB = \\begin{bmatrix} $(prod_11) & $(prod_12) \\\\ $(prod_21) & $(prod_22) \\end{bmatrix} = I\$",
                "Since \$AB = I\$, \$B = A^{-1}\$",
                sol("Answer", "Yes")
            ),
            time=150
        )
    end
end